Fixes #2300 Update langchain4j-jina module to support late chunking in long-context embedding models

ddobrin opened the pull request:
## Issue
Fixes #2300

## Change
Added support for Jina Embedding API v3, including support for late chunking".
Added tests to support previous version v2 and new version v3

## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [x] There are no breaking changes
- [x] I have added unit and/or integration tests for my change
- [x] The tests cover both positive and negative cases
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

ddobrin committed:
Fixes #2300 - Update langchain4j-jina module to support late chunking in long-context embedding models

ddobrin committed:
Fixes #2300 - Additional tests

ddobrin committed:
Merge branch 'langchain4j:main' into main

ddobrin committed:
Fixes #2300 - Switched Default Embedding model to jina-embeddings-v3

ddobrin committed:
Merge branch 'main' into main

ddobrin committed:
Merge branch 'main' into main

ddobrin committed:
Merge branch 'langchain4j:main' into main

ddobrin committed:
Fixes #2300 - Fixed github-actions bot recommendations

ddobrin committed:
Merge branch 'main' into main

ddobrin commented on langchain4j-jina/src/main/java/dev/langchain4j/model/jina/JinaEmbeddingModel.java:
```diff
@@ -31,24 +29,41 @@ public class JinaEmbeddingModel extends DimensionAwareEmbeddingModel {
     private final Integer maxRetries;
+    private final Boolean lateChunking;
+
+    @Builder
```
done

ddobrin commented on langchain4j-jina/src/main/java/dev/langchain4j/model/jina/JinaEmbeddingModel.java:
```diff
@@ -31,24 +29,41 @@ public class JinaEmbeddingModel extends DimensionAwareEmbeddingModel {
                 .baseUrl(getOrDefault(baseUrl, DEFAULT_BASE_URL))
                 .apiKey(apiKey)
                 .timeout(getOrDefault(timeout, ofSeconds(60)))
+                .lateChunking(getOrDefault(logRequests, false))
```
silly... corrected

ddobrin commented on langchain4j-jina/src/main/java/dev/langchain4j/model/jina/internal/client/JinaClient.java:
```diff
@@ -22,7 +21,13 @@ public class JinaClient {
+            String baseUrl,
+            String apiKey,
+            Duration timeout,
+            boolean lateChunking,
```
corrected

ddobrin commented on langchain4j-jina/src/main/java/dev/langchain4j/model/jina/JinaEmbeddingModel.java:
```diff
@@ -66,20 +81,17 @@ public Response<List<Embedding>> embedAll(List<TextSegment> textSegments) {
                 .model(modelName)
-                .input(textSegments.stream().map(TextSegment::text).collect(toList()))
+                .lateChunking(lateChunking)
+                .input(textSegments.stream().map(TextSegment::text).toList())
```
corrected

ddobrin commented on langchain4j-jina/src/main/java/dev/langchain4j/model/jina/JinaEmbeddingModel.java:
```diff
@@ -66,20 +81,17 @@ public Response<List<Embedding>> embedAll(List<TextSegment> textSegments) {
         List<Embedding> embeddings = response.data.stream()
                 .map(jinaEmbedding -> Embedding.from(jinaEmbedding.embedding))
-                .collect(toList());
+                .toList();
```
corrected

ddobrin commented on langchain4j-jina/src/test/java/dev/langchain4j/model/jina/JinaScoringModelIT.java:
```diff
@@ -63,9 +65,9 @@ void should_score_multiple_segments_with_all_parameters() {
         List<Double> scores = response.content();
         assertThat(scores).hasSize(2);
-        assertThat(scores.get(0)).isLessThan(scores.get(1));
+        assertThat(scores.get(1)).isLessThan(scores.get(0));
```
ReRanker has changed to jina-reranker-v2-base-multilingual - currently supported by Jina.
Fixed by adding separate tests for new and older jina-reranker-v1-turbo-en reranker

dliubarskyi replied:
perhaps it is a quirk of a new model, but from the test perspective it makes no sense that "maine coon" is more similar to "tell me about dogs" than "labrador retriever". So perhaps it makes sense to either leave a comment (to avoid confusion) or come up with another example that demonstrates that reranking actually works

ddobrin commented on langchain4j-jina/src/test/java/dev/langchain4j/model/jina/JinaEmbeddingModelIT.java:
```diff
@@ -1,25 +1,34 @@
         // given
-        EmbeddingModel model = JinaEmbeddingModel.withApiKey(System.getenv("JINA_API_KEY"));
+        EmbeddingModel model = JinaEmbeddingModel.builder()
+                .baseUrl("https://api.jina.ai/")
```
It is the latest version of the embedding model (v3)
previous model (v2) is older and does not support lateChunking

dliubarskyi replied:
I meant is there a reason to override base url, timeout, retires, etc? It adds a lot of noise, also not clear if those parameters are relevant to the test or no.

I would only set mandatory parameters and those parameters that are relevant to the test

dliubarskyi replied:
(same for all other added tests)

ddobrin commented on langchain4j-jina/src/test/java/dev/langchain4j/model/jina/JinaEmbeddingModelIT.java:
```diff
@@ -52,24 +94,166 @@ public void should_embed_multiple_segments() {
+                .modelName("jina-embeddings-v3")
+                .timeout(ofSeconds(10))
+                .maxRetries(2)
+                .lateChunking(true)
```
Illustrated in [link](https://jina.ai/news/jina-embeddings-v3-a-frontier-multilingual-embedding-model/#parameter-latechunking)

Excerpt:
From a user perspective, setting late_chunking does not change the input or output format. You will only notice a change in the embedding values, as they are now computed based on the entire previous context rather than independently.

We can observe the CosineSimilarity between embeddings changes as a result and is tested against 0.85, 0.6 vs 0.95, 0.9s

dliubarskyi replied:
I see, thanks!

Might want to do it in a single test:
- create model without late chunking
- calculate similarity
- create (same) model with late chunking
- calculate sililarity
- verify similarities are not the same (this verifies that setting `lateChunking` parameter does make an effect)

ddobrin committed:
Fixes #2300 - Fixed code review recommendations

ddobrin committed:
Merge remote-tracking branch 'origin/main'

ddobrin commented:
Hi @langchain4j 
Code review items resolved

ddobrin committed:
Merge branch 'main' into main

dliubarskyi commented on langchain4j-jina/src/test/java/dev/langchain4j/model/jina/JinaEmbeddingModelIT.java:
```diff
@@ -52,23 +94,165 @@ public void should_embed_multiple_segments() {
+    @Test
+    @DisplayName(
+            "Multiple text segments to embed, using Jina embedding model: jina-embeddings-v3; late chunking DISABLED")
+    void should_embed_multiple_segments_failure() {
```
what does `_failure` means here?

dliubarskyi reviewed (commented):
@ddobrin thank you!

ddobrin committed:
Fixes #2300 - Fixed code review recommendations for unit tests

ddobrin commented:
Thanks for the review @dliubarskyi 

- timeout makes sense to stay as the older v1 models are orders of magnitude slower; same latency observed in the jina.ai online embedding tester - it is possible that there is a cloud resource availability issue as it is inconsistent for the jina-reranker-v1-turbo-en model
- remove retries and baseURL though
- removed the failure test
- added an all-in-one comparison testL: should_embed_multiple_segments_compare_with_late_chunking()

dliubarskyi reviewed (approved):
@ddobrin thanks a lot!
